code/modules/power/lighting.dm

```diff
@@ -394,25 +406,53 @@
 	var/area/A = get_area(src)
 	return A.lightswitch && A.power_light
 
-/obj/machinery/light/proc/flicker(amount = rand(10, 20))
-	if(flickering)
+///flicker lights on and off
+/obj/machinery/light/proc/flicker(toggle_flicker = FALSE)
+	if(!has_power())
+		addtimer(CALLBACK(src, PROC_REF(flicker)), flicker_time)
 		return
-	flickering = TRUE
-	spawn(0)
-		if(light_on && status == LIGHT_OK)
-			for(var/i = 0; i < amount; i++)
-				if(status != LIGHT_OK)
-					break
-				update(FALSE)
-				sleep(rand(5, 15))
-			update(FALSE)
+	if(toggle_flicker)
+		if(status != LIGHT_OK)
+			return
+		flickering = flickering? FALSE : TRUE
+		if(flickering)
+			lightambient = new(null, FALSE)
+			lightambient.start(src)
+		else
+			lightambient.stop(src)
+	if(random_flicker)
+		flicker_time = rand(flicker_time_lower_min, flicker_time_upper_max)
+	if(status != LIGHT_OK)
+		lightambient.stop(src)
 		flickering = FALSE
+		return
+	light_flicker_state = light_flicker_state? FALSE : TRUE
+	if(light_flicker_state == FALSE)
+		if(base_icon_state == "tube")
+			flick("tube_flick_off", src)
+		else
+			flick("bulb_flick_off", src)
+		//delay the power change long enough to get the flick() animation off
+		addtimer(CALLBACK(src, PROC_REF(flicker_power_state)), 0.3 SECONDS)
+	else
+		if(base_icon_state == "tube")
+			flick("tube_flick_on", src)
+		else
+			flick("bulb_flick_on", src)
```

Comment:
normally we'd do something like "[base_icon_state]_flick_off" etc instead of this snowflake stuff.